### PR TITLE
Making substitute_array_list ignore out of bounds indexes

### PR DESCRIPTION
**Pull request by romainbrenguier**
When transforming array-list to (ARRAYOF/WITH) expression, we ignore
out of bounds indexes, there value are not meaningfull and may force
cbmc to allocate big arrays when trying to output concrete values.

This also adds unit test for this function.

### CODE REVIEW

File: src/solvers/refinement/string_refinement.cpp [hunk 1]
```diff
@@ -1679,18 +1682,17 @@ exprt string_refinementt::substitute_array_lists(exprt expr) const
...
     {
-      ret_expr=with_exprt(ret_expr,
-                          expr.operands()[i*2],
-                          expr.operands()[i*2+1]);
+      const exprt &index=expr.operands()[i*2];
+      const exprt &value=expr.operands()[i*2+1];
+      mp_integer index_value;
+      bool not_constant=to_integer(index, index_value);
+      if(not_constant ||
+         (index_value>=0 && index_value<=string_max_length))
```
Comment (allredj) [inline src/solvers/refinement/string_refinement.cpp]: `<string_max_length`

Comment (romainbrenguier) [inline src/solvers/refinement/string_refinement.cpp]: I updated that and the unit test to check for corner values
